touchforms/backend/touchcare.py

```diff
@@ -31,6 +36,171 @@
+        ref = instance.getReference()
+        def from_bundle(inst):
+            root = inst.getRoot()
+            root.setParent(instance.getBase())
+            return root
+
+        if 'casedb' in ref:
+            if settings.USES_SQLITE:
+                case_storage = self.sandbox.getCaseStorage()
+            else:
+                case_storage = CaseDatabase(
+                    self.vars.get('host'),
+                    self.vars['domain'],
+                    self.auth,
+                    self.vars.get("additional_filters", {}),
+                    self.vars.get("preload_cases", False),
+                    self.form_context,
+                )
+            return CaseInstanceTreeElement(
+                instance.getBase(),
+                case_storage,
+                False
+            )
+        elif 'fixture' in ref:
+            fixture_id = ref.split('/')[-1]
+            user_id = self.vars['user_id']
+            if settings.USES_SQLITE:
+                fixture = SqlSandboxUtils.loadFixture(self.sandbox, fixture_id, user_id)
+                root = fixture.getRoot()
+            else:
+                root = self._get_fixture(user_id, fixture_id)
+            root.setParent(instance.getBase())
+            return root
+        elif 'ledgerdb' in ref:
+            if settings.USES_SQLITE:
+                ledger_storage = self.sandbox.getLedgerStorage()
+            else:
+                ledger_storage = LedgerDatabase(
+                    self.vars.get('host'), self.vars['domain'],
+                    self.auth, self.vars.get("additional_filters", {}),
+                    self.vars.get("preload_cases", False),
+                )
+            return LedgerInstanceTreeElement(
+                instance.getBase(),
+                ledger_storage
+            )
+
+        elif 'session' in ref:
+            meta_keys = ['device_id', 'app_version', 'username', 'user_id']
+            exclude_keys = ['additional_filters', 'user_data']
+            sess = CommCareSession(None) # will not passing a CCPlatform cause problems later?
+            for k, v in self.vars.iteritems():
+                if k not in meta_keys \
+                        and k not in exclude_keys:
+                    # com.xhaus.jyson.JysonCodec returns data as byte strings
+                    # in unknown encoding (possibly ISO-8859-1)
+                    sess.setDatum(k, unicode(v, errors='replace'))
+
+            clean_user_data = {}
+            for k, v in self.vars.get('user_data', {}).iteritems():
+                clean_user_data[k] = unicode(v if v is not None else '', errors='replace')
+
+            return from_bundle(sess.getSessionInstance(*([self.vars.get(k, '') for k in meta_keys] + \
+                                                         [to_hashtable(clean_user_data)])))
+
+"""
+process_form_file and process_form_xml both perform submissions of the completed form form_data
+against the sandbox of the current user.
+
+"""
+
```

Comment:
extra space. usually this can go inside the function so it's easy to tell what the docstring is for